src/process/cli/sort.rs

```diff
@@ -0,0 +1,159 @@
+//! # Sort
+//!
+//! Sort rearranges the events so they are grouped by skb tracking id (or OVS queue_id if present)
+
+use std::{
+    fs::{File, OpenOptions},
+    io::{BufWriter, Write},
+    path::PathBuf,
+};
+
+use anyhow::{bail, Result};
+use clap::Parser;
+use log::debug;
+
+use crate::{
+    cli::*,
+    core::{
+        events::{file::FileEventsFactory, EventFactory, EventResult},
+        signals::Running,
+    },
+    module::Modules,
+    process::{
+        series::{EventSeries, EventSorter},
+        tracking::AddTracking,
+    },
+};
+
+/// The default size of the sorting buffer
+const DEFAULT_BUFFER: usize = 1000;
+
+/// Sort events in series based on tracking id.
+///
+/// Reads events from the INPUT file and arranges them by tracking id. The output is a number of
+/// "event sets". An event set is a list of events that share the same tracking id (i.e: belong to
+/// the same packet).
+#[derive(Parser, Debug, Default)]
+#[command(name = "sort")]
+pub(crate) struct Sort {
+    /// File from which to read events.
+    #[arg()]
+    pub(super) input: PathBuf,
+
+    /// Maximum number of events to buffer
+    ///
+    /// Sorting events requires storing events in a buffer while we wait to see if there is any
+    /// other event that belongs to the same series. If there are many interleaved events, you may
+    /// need to increase the size of the buffer to properly sort all events.
+    ///
+    /// A value of zero means the buffer can grow endlessly.
+    #[arg(long, default_value_t = DEFAULT_BUFFER)]
+    pub(super) max_buffer: usize,
+
+    /// Write event series to a file rather than to stdout.
+    #[arg(short, long)]
+    pub(super) out: Option<PathBuf>,
+
+    /// Write events to stdout even if --out is used.
+    #[arg(long, default_value = "false")]
+    pub(super) print: bool,
+
+    /// Json file writer if any.
+    #[arg(skip)]
+    file_writer: Option<BufWriter<File>>,
+}
+
+impl Sort {
+    // Print a series. For now we only support json format which is not very useful.
+    // FIXME: It's still unclear how the formatting API will look like. When it is, this should be
+    // adapted.
+    // For now, print a
+    fn print_series(series: &EventSeries) -> Result<()> {
+        println!("{}", &series.to_json());
+        Ok(())
+    }
+
+    fn write_json_series(&mut self, series: &EventSeries) -> Result<()> {
+        if let Some(writer) = self.file_writer.as_mut() {
+            let bytes = series.to_json().to_string().as_bytes().to_vec();
+            writer.write_all(&bytes)?;
+            writer.write_all(b"\n")?;
+            Ok(())
```

Comment:
I've tried, but the code ends up more ugly. IMHO, FormatAndWrite has proven not very useful. Sure, I can a Vect processor (which has to dump a json array not a list of json events) but initializing from  a list of abstract writers, where almost always we use 1, etc makes the code more ugly than rewriting the literally 3 line it provides.

Reply:
In addition, this is printing `Series`, which can hold more than just a list of events (e.g: we can calculate/print relative latencies), so the logic of how to format it (in this case into json) is currently in `Series::to_json()`.  Having `core::event::format` depend on an external thing so that we can add `FormatAndWrite::process_series()` is a bit odd.